### PR TITLE
Support Gradle configuration cache, small buildscript cleanups

### PR DESCRIPTION
**Pull request by dynpatch**
Hope this isn't too opinionated, feel free to ask for changes!

### CODE REVIEW

File: settings.gradle.kts [hunk 1]
```diff
@@ -16,3 +25,82 @@ pluginManagement {
...
+
+        maven("https://maven.su5ed.dev/releases") {
+            name = "Su5ed"
+        }
+    }
+}
+
+plugins {
+    id("org.gradle.toolchains.foojay-resolver-convention") version "0.9.0"
+}
```
Comment (BluSpring) [inline settings.gradle.kts]: From previous experience, using the Foojay Resolvers alongside Fabric Loom would always result in some kind of weird mixture of errors within the buildscript. I would avoid using this unless there is very good reason to.

File: build.gradle.kts [hunk 2]
```diff
@@ -229,16 +161,20 @@ java {
...
+    @get:Inject val fs: FileSystemOperations
+    @get:Inject val layout: ProjectLayout
+    @get:Inject val provider: ProviderFactory
+}
+
 tasks {
     register("countPatchProgress") {
         group = "kilt"
         description = "Counts the total of patches in Forge, and checks how many Kilt ForgeInjects there are, to check how much is remaining."
+        notCompatibleWithConfigurationCache("This isn't called every build; to be fixed later")
```
Comment (BluSpring) [inline build.gradle.kts]: The custom Kilt tasks are intentionally not called every build, they are used by me for determining how much I need to do in the codebase.

File: settings.gradle.kts [hunk 3]
```diff
@@ -16,3 +25,82 @@ pluginManagement {
...
+
+        maven("https://maven.terraformersmc.com/") {
+            name = "TerraformersMC"
+        }
+
+        maven("https://maven.su5ed.dev/releases") {
+            name = "Su5ed"
+        }
+    }
+}
```
Comment (BluSpring) [inline settings.gradle.kts:100]: Any specific reason as to why this has been moved here?

**Review by BluSpring — CHANGES_REQUESTED** [review]
Left a review, I'm curious to know about the reasoning behind these.

File: settings.gradle.kts (hunk 1 above)
Comment (qtdyn) [reply settings.gradle.kts]: Haven't had any issues in the past, but considering the nature of the project maybe it'd be best not to risk it

File: build.gradle.kts (hunk 2 above)
Comment (qtdyn) [reply build.gradle.kts]: Yup, maybe the "to be fixed later" phrasing wasn't useful, it's just to mark them as intentionnally not compatible to not trigger build failures

File: settings.gradle.kts (hunk 3 above)
Comment (qtdyn) [reply settings.gradle.kts:100]: That was just to unclutter the main buildscript and share them throughout subprojects (for potential modularization), this is mainly preference tho

File: settings.gradle.kts (hunk 1 above)
Comment (BluSpring) [reply settings.gradle.kts]: Please remove the plugin whenever you can.

File: build.gradle.kts (hunk 2 above)
Comment (BluSpring) [reply build.gradle.kts]: I would recommend changing the text to something else, maybe something along the lines of "Must not be cached".

File: settings.gradle.kts (hunk 3 above)
Comment (BluSpring) [reply settings.gradle.kts:100]: I'll allow it.

**Review by BluSpring — APPROVED** [review]

**Comment by BluSpring** [comment]
PR has been reverted due to there being a lot of issues with the buildscript, especially in launching the game. That one's on me, I didn't have time to check on it before merging, will actually spend time to try out the PRs next time.

**Comment by dynpatch** [comment]
I found an issue with repositories not being applied at times, which I've fixed in the new Gradle structure at #112 
Which other problems did you have? I'll try and remedy them as soon as i can, sorry for not having been as thorough for testing too.

**Comment by BluSpring** [comment]
There's an issue with launching the game itself, especially with Forge mods installed, where it would fail to properly apply the properties into the FMJ:

```log
[04:02:45] [main/INFO] (FabricLoader/GameProvider) Loading Minecraft 1.20.1 with Fabric Loader 0.16.9
[04:02:46] [ForkJoinPool-1-worker-2/WARN] (FabricLoader/Metadata) Mod not-that-cc uses the version ${version} which isn't compatible with Loader's extended semantic version format (Could not parse version number component '${version}'!), SemVer is recommended for reliably evaluating dependencies and prioritizing newer version
[04:02:46] [main/WARN] (FabricLoader/Resolution) Mod resolution failed
[04:02:46] [main/INFO] (FabricLoader/Resolution) Immediate reason: [HARD_DEP_NO_CANDIDATE kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends fabricloader @ [>=valueof(GradlePropertyValueSource)]}, ROOT_FORCELOAD_SINGLE kilt 0.0.1+mc1.20.1+build.7e7cb9-local]
[04:02:46] [main/INFO] (FabricLoader/Resolution) Reason: [HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends fabricloader @ [>=valueof(GradlePropertyValueSource)]}, HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends fabric-api @ [>=valueof(GradlePropertyValueSource)]}, HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends minecraft @ [valueof(GradlePropertyValueSource)]}, HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends fabric-language-kotlin @ [>=valueof(GradlePropertyValueSource)]}, HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends mm @ [>=valueof(GradlePropertyValueSource)]}, HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends forgeconfigapiport @ [>=valueof(GradlePropertyValueSource)]}, HARD_DEP kilt 0.0.1+mc1.20.1+build.7e7cb9-local {depends architectury @ [>=valueof(GradlePropertyValueSource)]}]
[04:02:46] [main/INFO] (FabricLoader/Resolution) Fix: add [], remove [], replace [[kilt 0.0.1+mc1.20.1+build.7e7cb9-local] -> add:kilt 1 ([(-∞,∞)])]
[04:02:46] [main/ERROR] (FabricLoader) Incompatible mods found!
net.fabricmc.loader.impl.FormattedException: Some of your mods are incompatible with the game or each other!
A potential solution has been determined, this may resolve your problem:
	 - Replace mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local with any version that is compatible with:
		 - fabricloader 0.16.9
		 - fabric-api 0.92.2+1.20.1
		 - minecraft 1.20.1
		 - fabric-language-kotlin 1.13.0+kotlin.2.1.0
		 - mm 2.3
		 - forgeconfigapiport 8.0.1
		 - architectury 9.2.14
More details:
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of mod 'Fabric Loader' (fabricloader), but only the wrong version is present: 0.16.9!
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of mod 'Fabric API' (fabric-api), but only the wrong version is present: 0.92.2+1.20.1!
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of 'Minecraft' (minecraft), but only the wrong version is present: 1.20.1!
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of mod 'Fabric Language Kotlin' (fabric-language-kotlin), but only the wrong version is present: 1.13.0+kotlin.2.1.0!
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of mod 'Manningham Mills' (mm), but only the wrong version is present: 2.3!
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of mod 'Forge Config API Port' (forgeconfigapiport), but only the wrong version is present: 8.0.1!
	 - Mod 'Kilt' (kilt) 0.0.1+mc1.20.1+build.7e7cb9-local requires version valueof(GradlePropertyValueSource) of mod 'Architectury' (architectury), but only the wrong version is present: 9.2.14!
	at net.fabricmc.loader.impl.FormattedException.ofLocalized(FormattedException.java:51) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.FabricLoaderImpl.load(FabricLoaderImpl.java:196) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:146) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:68) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.devlaunchinjector.Main.main(Main.java:86) ~[?:?]
```

Additionally, it seems like this PR somehow broke the DFU?
```log
[04:04:56] [main/ERROR] (FabricLoader) A mod crashed on startup!
net.fabricmc.loader.impl.FormattedException: java.lang.RuntimeException: Could not execute entrypoint stage 'preLaunch' due to errors, provided by 'kilt' at 'xyz.bluspring.kilt.loader.KiltEarlyInitializer'!
	at net.fabricmc.loader.impl.FormattedException.ofLocalized(FormattedException.java:63) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:162) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:68) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.devlaunchinjector.Main.main(Main.java:86) ~[dev-launch-injector-0.2.1+build.8.jar:?]
Caused by: java.lang.RuntimeException: Could not execute entrypoint stage 'preLaunch' due to errors, provided by 'kilt' at 'xyz.bluspring.kilt.loader.KiltEarlyInitializer'!
	at net.fabricmc.loader.impl.FabricLoaderImpl.lambda$invokeEntrypoints$2(FabricLoaderImpl.java:403) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.util.ExceptionUtil.gatherExceptions(ExceptionUtil.java:33) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:401) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:160) ~[fabric-loader-0.16.9.jar:?]
	... 3 more
Caused by: java.lang.ExceptionInInitializerError
	at net.minecraft.Util.doFetchChoiceType(Util.java:242) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.Util.fetchChoiceType(Util.java:235) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.entity.EntityType$Builder.build(EntityType.java:763) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.entity.EntityType.register(EntityType.java:168) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.entity.EntityType.<clinit>(EntityType.java:175) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.item.Items.<clinit>(Items.java:797) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.LightBlock.getShape(LightBlock.java:56) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.getShape(BlockBehaviour.java:1032) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.getShape(BlockBehaviour.java:1028) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour.getCollisionShape(BlockBehaviour.java:349) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase$Cache.<init>(BlockBehaviour.java:1293) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.initCache(BlockBehaviour.java:843) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.Blocks.<clinit>(Blocks.java:1185) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.FireBlock.bootStrap(FireBlock.java:326) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.server.Bootstrap.bootStrap(Bootstrap.java:54) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at xyz.bluspring.kilt.loader.KiltLoader.initForge(KiltLoader.kt:941) ~[main/:?]
	at xyz.bluspring.kilt.loader.KiltEarlyInitializer.onPreLaunch(KiltEarlyInitializer.kt:8) ~[main/:?]
	at net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:399) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:160) ~[fabric-loader-0.16.9.jar:?]
	... 3 more
Caused by: java.lang.NullPointerException: Cannot invoke "com.mojang.datafixers.types.templates.TypeTemplate.size()" because "this.f" is null
	at com.mojang.datafixers.types.templates.Product.size(Product.java:36) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.types.templates.Product.size(Product.java:36) ~[datafixerupper-6.0.8.jar:?]
	at java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:214) ~[?:?]
	at it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap$ValueSpliterator.acceptOnIndex(Object2ObjectOpenHashMap.java:1158) ~[fastutil-8.5.9.jar:?]
	at it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap$ValueSpliterator.acceptOnIndex(Object2ObjectOpenHashMap.java:1141) ~[fastutil-8.5.9.jar:?]
	at it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap$MapSpliterator.forEachRemaining(Object2ObjectOpenHashMap.java:814) ~[fastutil-8.5.9.jar:?]
	at it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap$ValueSpliterator.forEachRemaining(Object2ObjectOpenHashMap.java:1141) ~[fastutil-8.5.9.jar:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.IntPipeline.reduce(IntPipeline.java:520) ~[?:?]
	at java.util.stream.IntPipeline.max(IntPipeline.java:483) ~[?:?]
	at com.mojang.datafixers.types.templates.TaggedChoice.<init>(TaggedChoice.java:64) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.DSL.taggedChoice(DSL.java:195) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.DSL.taggedChoiceLazy(DSL.java:199) ~[datafixerupper-6.0.8.jar:?]
	at net.minecraft.util.datafix.schemas.V1460.method_5236(V1460.java:300) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at com.mojang.datafixers.schemas.Schema.resolveTemplate(Schema.java:91) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.schemas.Schema.getTemplate(Schema.java:103) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.schemas.Schema.buildTypes(Schema.java:47) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.schemas.Schema.<init>(Schema.java:38) ~[datafixerupper-6.0.8.jar:?]
	at net.minecraft.util.datafix.schemas.NamespacedSchema.<init>(NamespacedSchema.java:14) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at com.mojang.datafixers.DataFixerBuilder.addSchema(DataFixerBuilder.java:45) ~[datafixerupper-6.0.8.jar:?]
	at com.mojang.datafixers.DataFixerBuilder.addSchema(DataFixerBuilder.java:39) ~[datafixerupper-6.0.8.jar:?]
	at net.minecraft.util.datafix.DataFixers.addFixers(DataFixers.java:786) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.util.datafix.DataFixers.createFixerUpper(DataFixers.java:283) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.util.datafix.DataFixers.<clinit>(DataFixers.java:271) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.Util.doFetchChoiceType(Util.java:242) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.Util.fetchChoiceType(Util.java:235) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.entity.EntityType$Builder.build(EntityType.java:763) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.entity.EntityType.register(EntityType.java:168) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.entity.EntityType.<clinit>(EntityType.java:175) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.item.Items.<clinit>(Items.java:797) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.LightBlock.getShape(LightBlock.java:56) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.getShape(BlockBehaviour.java:1032) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.getShape(BlockBehaviour.java:1028) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour.getCollisionShape(BlockBehaviour.java:349) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase$Cache.<init>(BlockBehaviour.java:1293) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.initCache(BlockBehaviour.java:843) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.Blocks.<clinit>(Blocks.java:1185) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.world.level.block.FireBlock.bootStrap(FireBlock.java:326) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at net.minecraft.server.Bootstrap.bootStrap(Bootstrap.java:54) ~[minecraft-merged-f0654d1df1-1.20.1-loom.mappings.1_20_1.layered+hash.1899487486-v2.jar:?]
	at xyz.bluspring.kilt.loader.KiltLoader.initForge(KiltLoader.kt:941) ~[main/:?]
	at xyz.bluspring.kilt.loader.KiltEarlyInitializer.onPreLaunch(KiltEarlyInitializer.kt:8) ~[main/:?]
	at net.fabricmc.loader.impl.FabricLoaderImpl.invokeEntrypoints(FabricLoaderImpl.java:399) ~[fabric-loader-0.16.9.jar:?]
	at net.fabricmc.loader.impl.launch.knot.Knot.init(Knot.java:160) ~[fabric-loader-0.16.9.jar:?]
	... 3 more
```

**Comment by dynpatch** [comment]
The first issue seems relatively straightforward to fix, DFU on the other hand... what?

**Comment by BluSpring** [comment]
My thoughts exactly.

**Pull request merged**